### DISCUSSION TITLE
Cannot find libxxx needed by pygame and Pillow?

### DISCUSSION THREAD

**Discussion by pauljamescleary** [pyinstaller/pyinstaller#7343] [Help]
I have the mac version of the build working great with pyinstaller.  I am trying to get an ubuntu build working on Github Actions.

I am seeing some kind of issue finding dynamically linked dependencies and I am unclear on how to address.  I am hoping this is something obvious with pyinstaller + pygame + ubuntu that I haven't been able to sort out yet.

Here is the exact errror (there are lots of them):

```
19687 INFO: Looking for dynamic libraries
1839 WARNING: Cannot find libpng16-67f87fff.so.16.37.0 (needed by /opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/pygame/../pygame.libs/libSDL2_image-2-ca388ea9.0.so.0.2.3)
```

Here is a link to a build - https://github.com/pauljamescleary/chess/actions/runs/3749441152/jobs/6367924177#step:7:80

**Comment by bwoodsend** [comment]
I wouldn't worry about those. PyInstaller's warnings are rather naive. I've downloaded your linux build and inspected the contents with `pyi-archive_viewer` and those libraries are in there.

**Comment by rokm** [comment]
Those are warnings, not errors, and should not affect the build. 

The warnings happen because of how `rpaths` work on linux, and because PyInstaller naively and recursively analyses all collected binaries.

For example, in `site-packages/pygame`, you have extension called `_freetype.cpython-310-x86_64-linux-gnu.so` (replace 310 with corresponding python version). This extension sets its `rpath`:
```
$ objdump -p venv/lib/python3.10/site-packages/pygame/_freetype.cpython-310-x86_64-linux-gnu.so | egrep "RUNPATH|RPATH"
  RPATH                $ORIGIN/../pygame.libs
```

So if we analyze its imports, we resolve all pygame-bundled shared libs, because they are all resolvable with the given RPATH:
```
$ ldd venv/lib/python3.10/site-packages/pygame/_freetype.cpython-310-x86_64-linux-gnu.so
	linux-vdso.so.1 (0x00007ffdb5cc2000)
	libSDL2-2-302c83d0.0.so.0.16.0 => /home/rok/tmp/pyi-pygame/venv/lib/python3.10/site-packages/pygame/../pygame.libs/libSDL2-2-302c83d0.0.so.0.16.0 (0x00007f1bde639000)
	libfreetype-9936d4a5.so.6.18.0 => /home/rok/tmp/pyi-pygame/venv/lib/python3.10/site-packages/pygame/../pygame.libs/libfreetype-9936d4a5.so.6.18.0 (0x00007f1bde581000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f1bde558000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f1bde37b000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f1bde29b000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f1bde294000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f1bde28f000)
	libpng16-67f87fff.so.16.37.0 => /home/rok/tmp/pyi-pygame/venv/lib/python3.10/site-packages/pygame/../pygame.libs/libpng16-67f87fff.so.16.37.0 (0x00007f1bde258000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f1bde23e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f1bde7fb000)
```

Now, if we take a look at the bundled libfreetype shared library. It does not have an rpath set at all; running
```
$ objdump -p venv/lib/python3.10/site-packages/pygame.libs/libfreetype-9936d4a5.so.6.18.0 | egrep "RUNPATH|RPATH"
``` 
comes up empty. Therefore, when we analyze this shared library on its own, we cannot resolve the pygame-bundled shared libs (in this case, `libpng`):
``` 
$ ldd venv/lib/python3.10/site-packages/pygame.libs/libfreetype-9936d4a5.so.6.18.0
	linux-vdso.so.1 (0x00007ffd6fd44000)
	libpng16-67f87fff.so.16.37.0 => not found
	libm.so.6 => /lib64/libm.so.6 (0x00007f9777bfc000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f9777be2000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f9777bdd000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f9777a00000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9777dba000)
``` 

So this part of analysis (rightfully) generates a warning for `libpng16-67f87fff.so.16.37.0`, but the library is still collected because analysis of `freetype.cpython-310-x86_64-linux-gnu.so` did manage to pick it up (as rpath was set there and was correctly applied).
